dbms/src/IO/Compression/CompressionCodecRLE.cpp

```diff
@@ -0,0 +1,188 @@
+    UInt8 bytes_to_skip = (source_size - 1) % RLE_PAIR_LENGTH;
+    memcpy(dest, source, bytes_to_skip);
+    dest += bytes_to_skip;
+    source += bytes_to_skip;
+
+    while (source < source_end)
+    {
+        T data = unalignedLoad<T>(source);
+        source += sizeof(T);
+        auto count = unalignedLoad<UInt16>(source);
+        source += sizeof(UInt16);
+        if unlikely (dest + count * sizeof(T) > output_end)
+            throw Exception(
+                ErrorCodes::CANNOT_DECOMPRESS,
+                "Cannot decompress RLE-encoded data, output buffer is too small");
+        for (UInt16 i = 0; i < count; ++i)
+        {
+            unalignedStore<T>(dest, data);
+            dest += sizeof(T);
+        }
+    }
+}
+
+} // namespace
+
+UInt32 CompressionCodecRLE::doCompressData(const char * source, UInt32 source_size, char * dest) const
+{
+    switch (bytes_size)
+    {
+    case 1:
+        return compressDataForType<UInt8>(source, source_size, dest);
+    case 2:
+        return compressDataForType<UInt16>(source, source_size, dest);
+    case 4:
+        return compressDataForType<UInt32>(source, source_size, dest);
+    case 8:
+        return compressDataForType<UInt64>(source, source_size, dest);
+    default:
+        __builtin_unreachable();
+    }
+}
+
+void CompressionCodecRLE::doDecompressData(
+    const char * source,
+    UInt32 source_size,
+    char * dest,
+    UInt32 uncompressed_size) const
+{
+    if (source_size < 1)
+        throw Exception(ErrorCodes::CANNOT_DECOMPRESS, "Cannot decompress RLE-encoded data. File has wrong header");
+
+    if (uncompressed_size == 0)
+        return;
+
+    UInt8 bytes_size = source[0];
+    if (bytes_size == JUST_COPY_CODE)
+    {
+        if (source_size - 1 < uncompressed_size)
+            throw Exception(ErrorCodes::CANNOT_DECOMPRESS, "Cannot decompress RLE-encoded data. File has wrong header");
+
+        memcpy(dest, &source[1], uncompressed_size);
+        return;
+    }
+    if (bytes_size != 1 && bytes_size != 2 && bytes_size != 4 && bytes_size != 8)
+        throw Exception(ErrorCodes::CANNOT_DECOMPRESS, "Cannot decompress RLE-encoded data. File has wrong header");
+
+    UInt8 bytes_to_skip = uncompressed_size % (bytes_size + sizeof(UInt16));
+    if (static_cast<UInt32>(1 + bytes_to_skip) > source_size)
+        throw Exception(ErrorCodes::CANNOT_DECOMPRESS, "Cannot decompress RLE-encoded data. File has wrong header");
+
+    switch (bytes_size)
+    {
+    case 1:
+        decompressDataForType<UInt8>(source, source_size, dest, uncompressed_size);
+        break;
+    case 2:
+        decompressDataForType<UInt16>(source, source_size, dest, uncompressed_size);
+        break;
+    case 4:
+        decompressDataForType<UInt32>(source, source_size, dest, uncompressed_size);
+        break;
+    case 8:
+        decompressDataForType<UInt64>(source, source_size, dest, uncompressed_size);
+        break;
+    default:
+        __builtin_unreachable();
```

Comment:
If the file is corrupted, `bytes_size` could be a abnormal value.
I think we can throw here to provide more infomation?

Reply:
we have checked in line 162